### PR TITLE
Save Ping Instances preferences

### PR DESCRIPTION
**Pull request by tornac1234**
- [x] Save visibility preference
- [x] Save color preference

I also created a new InitialSyncProcessor which is meant to contain all the stuff that must occur only after everything has loaded

### CODE REVIEW

File: Nitrox.Test/Serialization/PingInstancePreferencesSerializationTest.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using NitroxModel.Logger;
+using NitroxServer.GameLogic.Players;
+using NitroxServer.Serialization;
+
+namespace NitroxTest.Serialization
+{
+    [TestClass]
+    public class PingInstancePreferencesSerializationTest
```
Comment (Jannify) [inline Nitrox.Test/Serialization/PingInstancePreferencesSerializationTest.cs:11]: Should be part of `WorldPersistenceTest.cs`

File: NitroxClient/GameLogic/InitialSync/PostInitialSyncProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+using NitroxClient.MonoBehaviours;
+using NitroxModel.Packets;
+using UnityEngine;
+
+namespace NitroxClient.GameLogic.InitialSync;
+
+/// <summary>
+/// Happens after each initial sync processor finished loading
+/// </summary>
+public class PostInitialSyncProcessor : InitialSyncProcessor
```
Comment (Jannify) [inline NitroxClient/GameLogic/InitialSync/PostInitialSyncProcessor.cs:16]: Imo there should not be a general `PostInitialSyncProcessor` for a concrete case. If the ping logic can't fit in existing `InitialSyncProcessor` then create a new one with concrete `DependentProcessors`.

File: NitroxModel/DataStructures/GameLogic/InitialPingInstancePreferences.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace NitroxModel.DataStructures.GameLogic
+{
+    [Serializable]
+    public class InitialPingInstancePreferences
+    {
+        public HashSet<string> HiddenSignalPings { get; set; }
+        public List<KeyValuePair<string, int>> ColorPreferences { get; set; }
```
Comment (Jannify) [inline NitroxModel/DataStructures/GameLogic/InitialPingInstancePreferences.cs:10]: We discussed that Dictionary didn't worked initially. Did the provided fix not work? I would like to have a `Dictionary<>`, `IDictionary<>`, or `ThreadSafeDictionary` here more than `List<KeyValuePair<>>`

File: NitroxPatcher/Patches/Dynamic/PingInstance_SetVisible_Patch.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    public static void Prefix(PingInstance __instance, bool visible)
+    {
+        // Only notice server if there's a change on client-side
+        if (__instance.visible == visible)
+        {
+            return;
+        }
+
+        if (PingInstance_SetColor_Patch.TryGetPingKey(__instance.gameObject, out string pingKey))
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/PingInstance_SetVisible_Patch.cs:24]: In discord you asked if you can have two patches in one file. What was your final reasoning against that?

File: NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public override Version TargetVersion { get; } = new Version(1, 7, 0, 0);
+
+        protected override void UpgradePlayerData(JObject data)
+        {
+            foreach (JToken playerEntry in data["Players"])
+            {
+                playerEntry["HiddenSignalPings"] = new JArray();
+            }
+        }
```
Comment (Jannify) [inline NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs:17]: This should not work as the new field in [PersistedPlayerData.cs](https://github.com/SubnauticaNitrox/Nitrox/pull/1810/files#diff-6fd93b15a768ba981f6f0524d52a22d69d3ea35dac8994a320f1689bad2a4028) is called `PingInstancePreferences`, or am I wrong?

File: NitroxServer/Player.cs [hunk 6]
```diff
@@ -172,5 +175,29 @@ protected bool Equals(Player other)
...
+                PingInstancePreferences.HiddenSignalPings.Remove(pingKey);
+                return;
+            }
+
+            PingInstancePreferences.HiddenSignalPings.Add(pingKey);
+        }
+
+        public void SetPingColor(string pingKey, int color)
+        {
+            // Default color which we don't want to notice
```
Comment (Jannify) [inline NitroxServer/Player.cs:193]: What is the reason we don't want to notice when color is 0?

File: NitroxServer/GameLogic/Players/PingInstancePreferences.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        HiddenSignalPings = hiddenSignalPings;
+        ColorPreferences = colorPreferences;
+    }
+
+    public override string ToString()
+    {
+        return $"[HiddenSignalPings: {HiddenSignalPings.Count}, colorPreferences: {ColorPreferences.Count}]";
+    }
+
```
Comment (Jannify) [inline NitroxServer/GameLogic/Players/PingInstancePreferences.cs:34]:
```suggestion
```

File: NitroxPatcher/Patches/Dynamic/PingInstance_SetColor_Patch.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            else
+            {
+                Log.Warn($"Couldn't find PingInstance identifier for {gameObject.name}");
+            }
+        }
+        else
+        {
+            pingKey = entity.Id.ToString();
+        }
+        return pingKey != string.Empty;
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/PingInstance_SetColor_Patch.cs:52]: Didn't test this ingame:
```suggestion
        if (NitroxEntity.TryGetEntityFrom(gameObject, out NitroxEntity entity))
        {
            pingKey = entity.Id.ToString();
            return true; // Id of a NitroxEntity should never be null.
        }
        
        if (gameObject.TryGetComponent(out SignalPing signalPing))
        {
            pingKey = signalPing.descriptionKey;
            return !string.IsNullOrEmpty(pingKey);
        }
        
        Log.Warn($"Couldn't find PingInstance identifier for {gameObject.name}");
        pingKey = string.Empty;
        return false;
```

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs (hunk 5 above)
Comment (tornac1234) [reply NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs:17]: yeah I forgot to delete that file

File: NitroxServer/Player.cs (hunk 6 above)
Comment (tornac1234) [reply NitroxServer/Player.cs:193]: because it will be set to 0 by default on client-side so it's just saving memory

Comment (Jannify) [reply NitroxServer/Player.cs:193]: I would say that the memory you get here is not worth the complexity. I guess it's one of the reasons you don't use a Dictionary with a custom class or Tuple to save visibility and color. Imo you should switch to that.

File: NitroxModel/DataStructures/GameLogic/InitialPingInstancePreferences.cs (hunk 3 above)
Comment (tornac1234) [reply NitroxModel/DataStructures/GameLogic/InitialPingInstancePreferences.cs:10]: it seems to work fine without the fix

File: Nitrox.Test/Serialization/WorldPersistenceTest.cs [hunk 9]
```diff
@@ -331,6 +330,12 @@ public void PlayerDataTest()
                     Assert.AreEqual(playerData.Permissions, playerDataAfter.Permissions, $"PlayerData.Players.Permissions is not equal while using {serverSerializers[serializerIndex]}.");
                     Assert.AreEqual(playerData.NitroxId, playerDataAfter.NitroxId, $"PlayerData.Players.NitroxId is not equal while using {serverSerializers[serializerIndex]}.");
                     Assert.AreEqual(playerData.IsPermaDeath, playerDataAfter.IsPermaDeath, $"PlayerData.Players.IsFurniture is not equal while using {serverSerializers[serializerIndex]}.");
+
+                    foreach (string pingKey in playerData.PingInstancePreferences.Keys)
+                    {
+                        Assert.AreEqual(playerDataAfter.PingInstancePreferences[pingKey], playerDataAfter.PingInstancePreferences[pingKey], $"PlayerData.Players.PingInstancePreferences is not equal while using {serverSerializers[serializerIndex]}.");
```
Comment (tornac1234) [inline Nitrox.Test/Serialization/WorldPersistenceTest.cs:336]: will also correct that

File: NitroxModel/DataStructures/GameLogic/PingInstancePreference.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        return obj is PingInstancePreference otherInstance && Color == otherInstance.Color && Visible == otherInstance.Visible;
+    }
+
+    public override int GetHashCode()
+    {
+        int hashCode = 295593209;
+        hashCode = hashCode * -1521134295 + Color.GetHashCode();
+        hashCode = hashCode * -1521134295 + Visible.GetHashCode();
+        return hashCode;
+    }
```
Comment (Jannify) [inline NitroxModel/DataStructures/GameLogic/PingInstancePreference.cs:37]: iirc we decided that we won't implement equality functions if they are only used for the world persistent test.

File: Nitrox.Test/Serialization/WorldPersistenceTest.cs [hunk 11]
```diff
@@ -331,6 +330,12 @@ public void PlayerDataTest()
                     Assert.AreEqual(playerData.Permissions, playerDataAfter.Permissions, $"PlayerData.Players.Permissions is not equal while using {serverSerializers[serializerIndex]}.");
                     Assert.AreEqual(playerData.NitroxId, playerDataAfter.NitroxId, $"PlayerData.Players.NitroxId is not equal while using {serverSerializers[serializerIndex]}.");
                     Assert.AreEqual(playerData.IsPermaDeath, playerDataAfter.IsPermaDeath, $"PlayerData.Players.IsFurniture is not equal while using {serverSerializers[serializerIndex]}.");
+
+                    foreach (string pingKey in playerData.PingInstancePreferences.Keys)
+                    {
+                        Assert.AreEqual(playerData.PingInstancePreferences[pingKey], playerDataAfter.PingInstancePreferences[pingKey], $"PlayerData.Players.PingInstancePreferences is not equal while using {serverSerializers[serializerIndex]}.");
+                    }
+                    Assert.AreEqual(playerData.PingInstancePreferences.Count, playerDataAfter.PingInstancePreferences.Count, $"PlayerData.Players.PingInstancePreferences is not equal while using {serverSerializers[serializerIndex]}.");
```
Comment (Jannify) [inline Nitrox.Test/Serialization/WorldPersistenceTest.cs:338]: Count should be before the for loop.

File: NitroxServer/Player.cs [hunk 12]
```diff
@@ -33,12 +34,13 @@ public class Player : IProcessorContext
...
         public Optional<NitroxId> LastStoredSubRootID { get; set; }
         public ThreadSafeSet<string> CompletedGoals { get; }
+        public Dictionary<string, PingInstancePreference> PingInstancePreferences { get; set; }
         public bool IsMuted { get; set; }
 
         public Player(ushort id, string name, bool isPermaDeath, PlayerContext playerContext, NitroxConnection connection,
                       NitroxVector3 position, NitroxId playerId, Optional<NitroxId> subRootId, Perms perms, PlayerStatsData stats,
                       IEnumerable<NitroxTechType> usedItems, IEnumerable<string> quickSlotsBinding,
-                      IEnumerable<EquippedItemData> equippedItems, IEnumerable<EquippedItemData> modules, HashSet<string> completedGoals)
+                      IEnumerable<EquippedItemData> equippedItems, IEnumerable<EquippedItemData> modules, HashSet<string> completedGoals, Dictionary<string, PingInstancePreference> pingInstancePreferences)
```
Comment (Jannify) [inline NitroxServer/Player.cs:43]: Should be IDictionary cast to an ThreadSafeDictionary

File: NitroxPatcher/Patches/Dynamic/PingInstance_Set_Patches.cs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+    public static void PrefixVisible(PingInstance __instance, bool visible)
+    {
+        // Only notice server if there's a change on client-side
+        if (__instance.visible == visible)
+        {
+            return;
+        }
+
+        if (PlayerPreferencesInitialSyncProcessor.TryGetKeyForPingInstance(__instance, out string pingKey, out bool isRemotePlayerPing))
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/PingInstance_Set_Patches.cs:40]:
```suggestion
        if (PlayerPreferencesInitialSyncProcessor.TryGetKeyForPingInstance(__instance, out string pingKey, out bool _))
```

File: NitroxPatcher/Patches/Dynamic/PingInstance_Set_Patches.cs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+    public static void PrefixColor(PingInstance __instance, int index)
+    {
+        // Only notice server if there's a change on client-side
+        if (__instance.colorIndex == index)
+        {
+            return;
+        }
+        
+        if (PlayerPreferencesInitialSyncProcessor.TryGetKeyForPingInstance(__instance, out string pingKey, out bool isRemotePlayerPing))
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/PingInstance_Set_Patches.cs:26]:
```suggestion
        if (PlayerPreferencesInitialSyncProcessor.TryGetKeyForPingInstance(__instance, out string pingKey, out bool _))
```

File: NitroxPatcher/Patches/Dynamic/PingInstance_Set_Patches.cs [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+        if (PlayerPreferencesInitialSyncProcessor.TryGetKeyForPingInstance(__instance, out string pingKey, out bool isRemotePlayerPing))
+        {
+            UpdatePingInstancePreference(pingKey, visible, __instance.colorIndex);
+        }
+    }
+
+    private static void UpdatePingInstancePreference(string pingKey, bool visible, int color)
+    {
+        Resolve<IPacketSender>().Send(new SignalPingPreferenceChanged(pingKey, visible, color));
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/PingInstance_Set_Patches.cs:48]: Imo we shouldn't have a dedicated method for only one line.

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxModel/DataStructures/GameLogic/PingInstancePreference.cs (hunk 10 above)
Comment (tornac1234) [reply NitroxModel/DataStructures/GameLogic/PingInstancePreference.cs:37]: So should i just move the equality function to the test class ?

Comment (Jannify) [reply NitroxModel/DataStructures/GameLogic/PingInstancePreference.cs:37]: Yeah, like it does in the the rest of the class

File: NitroxClient/GameLogic/InitialSync/PlayerPreferencesInitialSyncProcessor.cs [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+        GameObject.FindObjectsOfType<PingInstance>().ForEach(updateInstance);
+    }
+
+    /// <summary>
+    /// Updateds the given pingInstance if it has a specified preference
+    /// </summary>
+    private void ModifyPingInstanceIfPossible(PingInstance pingInstance, Dictionary<string, PingInstancePreference> preferences, Action callback)
+    {
+        if (TryGetKeyForPingInstance(pingInstance, out string pingKey, out bool isRemotePlayerPing, callback)
+            && preferences.TryGetValue(pingKey, out PingInstancePreference preference))
```
Comment (Jannify) [inline NitroxClient/GameLogic/InitialSync/PlayerPreferencesInitialSyncProcessor.cs:59]:
```suggestion
        if (TryGetKeyForPingInstance(pingInstance, out string pingKey, out bool isRemotePlayerPing, callback) &&
            preferences.TryGetValue(pingKey, out PingInstancePreference preference))
```

File: NitroxClient/GameLogic/InitialSync/PlayerPreferencesInitialSyncProcessor.cs [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+    // Right after initial sync modifications, uGUI_PingEntry elements don't show their updated state
+    private void RefreshPingEntryInPDA(PingInstance pingInstance)
+    {
+        uGUI_PingTab pingTab;
+        if (!uGUI_PDA.main || !uGUI_PDA.main.tabs.TryGetValue(PDATab.Ping, out uGUI_PDATab pdaTab))
+        {
+            return;
+        }
+        pingTab = pdaTab as uGUI_PingTab;
```
Comment (Jannify) [inline NitroxClient/GameLogic/InitialSync/PlayerPreferencesInitialSyncProcessor.cs:81]:
```suggestion
        if (!uGUI_PDA.main || !uGUI_PDA.main.tabs.TryGetValue(PDATab.Ping, out uGUI_PDATab pdaTab))
        {
            return;
        }

        uGUI_PingTab pingTab = pdaTab as uGUI_PingTab;
```

File: NitroxClient/GameLogic/InitialSync/PlayerPreferencesInitialSyncProcessor.cs [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+        if (pingTab && pingTab.entries.TryGetValue(pingInstance.GetInstanceID(), out uGUI_PingEntry pingEntry))
+        {
+            pingEntry.SetColor(pingInstance.colorIndex);
+            pingEntry.SetVisible(pingInstance.visible);
+        }
+    }
+
+    // We need to differentiate two types of pings, the "normal pings" from objects that emit a signal, these objects generally contain a NitroxEntity
+    // Another type is Signal pings that are generated by the story events, they are located in the Global Root and don't contain a NitroxEntity, to be identified, they have another object: a SignalPing which contains a description key
+    // The last type possible is RemotePlayers' pings which are located in a GameObject that is 2 steps under the main object
```
Comment (Jannify) [inline NitroxClient/GameLogic/InitialSync/PlayerPreferencesInitialSyncProcessor.cs:91]:
```suggestion
    // We need to differentiate three types of pings, the "normal pings" from objects that emit a signal, these objects generally contain a NitroxEntity
    // Another type is Signal pings that are generated by the story events, they are located in the Global Root and don't contain a NitroxEntity, to be identified, they have another object: a SignalPing which contains a description key
    // The last type possible is RemotePlayers' pings which are located in a GameObject that is 2 steps under the main object
```

**Review by Jannify — CHANGES_REQUESTED** [review]

**Review by Jannify — APPROVED** [review]
Looks fine code wise. Needs a merge from master and in-game testing.

**Comment by tornac1234** [comment]
merge when 🥱

**Comment by Crapy** [comment]
Tested this and it works fine in-game

**Pull request merged**